Extract abstracts from Pubmed articles, even if they have no extra label

FlorianH5 opened the pull request:
### Description
This pull request involves modifications to the extraction method for abstracts/summaries within the PubMed utility. A condition has been added to verify the presence of unlabeled abstracts. Now an abstract will be extracted even if it does not have a subtitle. In addition, the extraction of the abstract was extended to books.

### Issue
The PubMed utility occasionally returns an empty result when extracting abstracts from articles, despite the presence of an abstract for the paper on PubMed. This issue arises due to the varying structure of articles; some articles follow a "subtitle/label: text" format, while others do not include subtitles in their abstracts. An example of the latter case can be found at: [https://pubmed.ncbi.nlm.nih.gov/37666905/](url)

<!-- Thank you for contributing to LangChain!

Replace this entire comment with:
  - Description: a description of the change, 
  - Issue: the issue # it fixes (if applicable),
  - Dependencies: any dependencies required for this change,
  - Tag maintainer: for a quicker response, tag the relevant maintainer (see below),
  - Twitter handle: we announce bigger features on Twitter. If your PR gets announced and you'd like a mention, we'll gladly shout you out!

Please make sure your PR is passing linting and testing before submitting. Run `make format`, `make lint` and `make test` to check this locally.

See contribution guidelines for more information on how to write/run tests, lint, etc: 
https://github.com/hwchase17/langchain/blob/master/.github/CONTRIBUTING.md

If you're adding a new integration, please include:
  1. a test for the integration, preferably unit tests that do not rely on network access,
  2. an example notebook showing its use. These live is docs/extras directory.

If no one reviews your PR within a few days, please @-mention one of @baskaryan, @eyurtsev, @hwchase17, @rlancemartin.
 -->

FlorianH5 committed:
Extract abstract even if it has no label

baskaryan commented on libs/langchain/langchain/utilities/pubmed.py:
```diff
@@ -161,13 +161,13 @@ def _parse_article(self, uid: str, text_dict: dict) -> dict:
+            for txt in abstract_text
+            if "#text" in txt and "@Label" in txt
+        ]
+        summary = "\n".join(summaries) if summaries else abstract_text
```
should abstract_text be converted to a string?

FlorianH5 replied:
This does not seem to be necessary, since the abstract is either list of dictionaries in the form "label: text" or a string.

However, to account for possible special cases, abstract_text could be checked as followed:
`summary = "\n".join(summaries) if summaries else abstract_text if isinstance(abstract_text, str) else "\n".join(str(item) for item in abstract_text)`

FlorianH5 replied:
After further testing, I found another problem that should be fixed in the latest commit. Besides the problem with extracting the abstracts of articles, the current implementation also throws a KeyError when a book is given as input. So I added a handler for this case. In addition, I've updated the "summary" part accordingly.

An example for the KeyError case is this book: [https://pubmed.ncbi.nlm.nih.gov/29262144/](url)

FlorianH5 committed:
Extended parsing to abstracts from books

Without these changes, the _parse_article method throws a KeyError when a book is given as input.

FlorianH5 committed:
Update pubmed.py so that an empty abstract can be handled

eyurtsev commented on libs/langchain/langchain/utilities/pubmed.py:
```diff
@@ -158,15 +158,30 @@ def retrieve_article(self, uid: str, webenv: str) -> dict:
-                f"{txt['@Label']}: {txt['#text']}"
-                for txt in ar.get("Abstract", {}).get("AbstractText", [])
-                if "#text" in txt and "@Label" in txt
+        try:
```
Could you add a test here: 

https://github.com/langchain-ai/langchain/blob/eugene%2Fautomaton_variant_3/libs/langchain/tests/integration_tests/utilities/test_pubmed.py#L13-L13


Ideally, we would have a test fixture for the actual parsing logic

cc @leo-gan

eyurtsev reviewed (commented):
Thanks so much for the contribution! Let me know if you're able to add a unit test or expand the existing integration test -- it'll help bullet proof the code against future modifications

FlorianH5 commented:
> Thanks so much for the contribution! Let me know if you're able to add a unit test or expand the existing integration test -- it'll help bullet proof the code against future modifications

Thank you very much for letting me contribute to the project. I would like to extend the existing integration test.

FlorianH5 committed:
Updated the integration test for pubmed.py

FlorianH5 commented:
> > Thanks so much for the contribution! Let me know if you're able to add a unit test or expand the existing integration test -- it'll help bullet proof the code against future modifications
> 
> Thank you very much for letting me contribute to the project. I would like to extend the existing integration test.

I have updated the existing integration test. @eyurtsev @baskaryan

FlorianH5 committed:
Make format

baskaryan committed:
Merge branch 'master' into FlorianH5/master

baskaryan commented:
apologies for the delay, thank you @FlorianH5!
